Sort map by key before printing

I wish there is an option / set of methods that will print maps with keys sorted in a best-effort manner. This can be extremely helpful in diffing outputs across runs.

(As you likely know, spew has a `SortKeys` global config that achieves exactly this.)

Thank you for your consideration.

> Maybe be we could create a new branch, which branch named v3. if we want to get lower version code, we can do it below.

Done. The following command seems to work, although I think it would require Go 1.11:

```
$ GO111MODULE=on go get github.com/k0kubun/pp/v3
go: extracting github.com/k0kubun/pp/v3 v3.0.5
go: extracting github.com/k0kubun/pp v3.0.1+incompatible
```

I still have no idea why `v3.0.1+incompatible` is shown while v3 branch and v3.0.5 tag seem to support go module.

> We can use internal/fmtsort package to make the key order same as the fmt package

Looks like this is not quite true:

```
printer.go:6:2: use of internal package internal/fmtsort not allowed
```

--- a/printer.go
+++ b/printer.go
@@ -170,10 +170,9 @@ func (p *printer) printMap() {
 		p.println("{")
 	}
 	p.indented(func() {
-		keys := p.value.MapKeys()
-		for i := 0; i < p.value.Len(); i++ {
-			value := p.value.MapIndex(keys[i])
-			p.indentPrintf("%s:\t%s,\n", p.format(keys[i]), p.format(value))
+		value := sortMap(p.value)
+		for i := 0; i < value.Len(); i++ {
+			p.indentPrintf("%s:\t%s,\n", p.format(value.keys[i]), p.format(value.values[i]))
 		}
 	})
 	p.indentPrint("}")

--- a/printer_test.go
+++ b/printer_test.go
@@ -178,17 +178,16 @@ var (
 			}
 			`,
 		},
-		// Note: this seems like failing randomly
-		// {
-		// 	map[string]interface{}{"foo": 10, "bar": map[int]int{20: 30}}, `
-		// 	[green]map[string]interface {}[reset]{
-		// 	  [red][bold]"[reset][red]foo[reset][red][bold]"[reset]: [blue][bold]10[reset],
-		// 	  [red][bold]"[reset][red]bar[reset][red][bold]"[reset]: [green]map[int]int[reset]{
-		// 	    [blue][bold]20[reset]: [blue][bold]30[reset],
-		// 	  },
-		// 	}
-		// 	`,
-		// },
+		{
+			map[string]interface{}{"foo": 10, "bar": map[int]int{20: 30}}, `
+			[green]map[string]interface {}[reset]{
+			  [red][bold]"[reset][red]bar[reset][red][bold]"[reset]: [green]map[int]int[reset]{
+			    [blue][bold]20[reset]: [blue][bold]30[reset],
+			  },
+			  [red][bold]"[reset][red]foo[reset][red][bold]"[reset]: [blue][bold]10[reset],
+			}
+			`,
+		},
 	}
 
 	arr [3]int

--- a/sort.go
+++ b/sort.go
@@ -1,0 +1,75 @@
+package pp
+
+import (
+	"reflect"
+	"sort"
+)
+
+func sortMap(value reflect.Value) *sortedMap {
+	if value.Type().Kind() != reflect.Map {
+		panic("sortMap is used for a non-Map value")
+	}
+
+	keys := make([]reflect.Value, 0, value.Len())
+	values := make([]reflect.Value, 0, value.Len())
+	mapRange := value.MapRange()
+	for mapRange.Next() {
+		keys = append(keys, mapRange.Key())
+		values = append(values, mapRange.Value())
+	}
+
+	sorted := &sortedMap{
+		keys:   keys,
+		values: values,
+	}
+	sort.Stable(sorted)
+	return sorted
+}
+
+type sortedMap struct {
+	keys   []reflect.Value
+	values []reflect.Value
+}
+
+// Functions for sort.Interface
+
+func (s *sortedMap) Len() int {
+	return len(s.keys)
+}
+
+func (s *sortedMap) Swap(i, j int) {
+	s.keys[i], s.keys[j] = s.keys[j], s.keys[i]
+	s.values[i], s.values[j] = s.values[j], s.values[i]
+}
+
+func (s *sortedMap) Less(i, j int) bool {
+	a, b := s.keys[i], s.keys[j]
+	if a.Type() != b.Type() {
+		return false // give up
+	}
+
+	// Return true if b is bigger
+	switch a.Kind() {
+	case reflect.Int, reflect.Int8, reflect.Int16, reflect.Int32, reflect.Int64:
+		return a.Int() < b.Int()
+	case reflect.Uint, reflect.Uint8, reflect.Uint16, reflect.Uint32, reflect.Uint64, reflect.Uintptr:
+		return a.Uint() < b.Uint()
+	case reflect.String:
+		return a.String() < b.String()
+	case reflect.Float32, reflect.Float64:
+		if a.Float() != a.Float() || b.Float() != b.Float() {
+			return false // NaN
+		}
+		return a.Float() < b.Float()
+	case reflect.Bool:
+		return !a.Bool() && b.Bool()
+	case reflect.Ptr:
+		return a.Pointer() < b.Pointer()
+	case reflect.Struct:
+		return a.NumField() < b.NumField()
+	case reflect.Array:
+		return a.Len() < b.Len()
+	default:
+		return false // not supported yet
+	}
+}

